### PR TITLE
Add gitlab-event function

### PR DESCRIPTION
**Pull request by martindekov**
## Description

This function is the front-facing function which get events from GitLab

Conceptual:

asdzxc

## How Has This Been Tested?

Have `gitlab.yml` file looking like 
```
provider:
  name: faas
  gateway: http://127.0.0.1:8080

functions:
  system-gitlab-event:
    lang: go
    handler: ./gitlab-event
    image: martindekov/gitlab-event:0.0.1
    labels:
      openfaas-cloud: "1"
      role: openfaas-system
    environment:
      content_type: text/plain
      validate_customers: true
      validate_token: true
      write_debug: true
      read_debug: true
      gitlab_instance: "https://gitlab.dedal.io/"
    environment_file:
      - gateway_config.yml
    secrets:
      - gitlab-webhook-secret
      - payload-secret
      - gitlab-api-token
```

Where secrets are:
 
* `gitlab-webhook-secret` is the secret provided when creating `System Hook`

* `payload-secret` is the internal trust secret

* `gitlab-api-token` is the api Access token to check if `openfaas-cloud` tag is present which means the project is part of the openfaas-cloud "installation"

And env vars are:

* `validate_customers: true` for customer of the instance validation

*  `validate_token: true` to match `gitlab-webhook-secret` with the secret sent with the `System Hook`

* `gitlab_instance: "https://gitlab.dedal.io/"` the base URL of your GitLab instance since the delete/update events don't send the URL

## How are existing users impacted? What migration steps/scripts do we need?

They will have GitLab front-facing function which triggers events

## Checklist:

I have:

- [x] updated the documentation and/or roadmap (if required)
- [x] read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] signed-off my commits with `git commit -s`
- [x] added unit tests

### CODE REVIEW

File: gitlab-event/handler.go [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+	}
+	projectInfo := GitLabProjectTags{}
+
+	unmarshalErr := json.Unmarshal(body, &projectInfo)
+	if unmarshalErr != nil {
+		return false, unmarshalErr
+	}
+
+	for _, tag := range projectInfo.TagList {
+		if strings.EqualFold(tag, "openfaas-cloud") {
```
Comment (WTFKr0) [inline gitlab-event/handler.go]: Can we make this tag in env var too ?
With a default to openfaas-cloud
If this solution match our needs, we probably have multiple openfaas-cloud isolated envs, and user should choose on which deploy his function
Like faas-dev and faas-prod for instance

Comment (martindekov) [reply gitlab-event/handler.go]: Very good idea! Thank you for the suggestion @WTFKr0 I will add that

File: gitlab-event/handler.go [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+			return fmt.Sprintf("Error while trying to connect to GitLab API: %s", err.Error())
+		}
+		if installed {
+			headers := map[string]string{
+				"X-Gitlab-Token": xGitlabToken,
+				"X-Gitlab-Event": eventHeader,
+				"Content-Type":   "application/json",
+			}
+
+			body, statusCode, err := forward(req, "gitlab-push", headers)
```
Comment (bartsmykla) [inline gitlab-event/handler.go]: I think that you should check error first then compare status codes.

File: gitlab-event/handler.go [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+				return err.Error()
+			}
+
+			return body
+		}
+		return fmt.Sprintf("To install project on the openfaas-cloud instance add \"%s\" tag", installationTag)
+
+	case "project_update", "project_destroy":
+		eventInfo := GitLabProjectEvent{}
+		json.Unmarshal(req, &eventInfo)
```
Comment (bartsmykla) [inline gitlab-event/handler.go]: I think that you should check error here

File: gitlab-event/handler.go [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+			found = true
+			break
+		}
+	}
+	return found
+}
+
+func getUser(pathWithNamespace string) string {
+	return pathWithNamespace[:strings.Index(pathWithNamespace, "/")]
+}
```
Comment (bartsmykla) [inline gitlab-event/handler.go]: I think that this function should be unit tested, what will happen when username will be in different format that expected

File: gitlab.yml [hunk 5]
```diff
@@ -0,0 +1,26 @@
+provider:
+  name: faas
+  gateway: http://127.0.0.1:8080
+
+functions:
+  system-gitlab-event:
+    lang: go
+    handler: ./gitlab-event
+    image: martindekov/gitlab-event:0.0.1
```
Comment (bartsmykla) [inline gitlab.yml]: I think that version should be `0.1.0`

File: gitlab.yml [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+      openfaas-cloud: "1"
+      role: openfaas-system
+    environment:
+      content_type: text/plain
+      validate_customers: true
+      validate_token: false
+      write_debug: true
+      read_debug: true
+      installation_tag: "openfaas-cloud"
+      gitlab_instance: "https://gitlab.dedal.io/"
```
Comment (bartsmykla) [inline gitlab.yml]: I think that you should put here some example url not this one, because it can be not valid in the future

File: docs/README.md [hunk 7]
```diff
@@ -650,3 +650,158 @@ You will need to distribute or share pub-cert.pem so that people can use this wi
 ### Wildcard domains with of-router
 
 Coming soon
+
+# GitLab
+
+Currently the the only way you can get your OpenFaas Cloud working with GitLab is on-prem.
```
Comment (bartsmykla) [inline docs/README.md]: @ivanayov suggested to change that sentence. There is also duplicated word `the`

File: gitlab.yml (hunk 5 above)
Comment (alexellis) [reply gitlab.yml]: Sure, and we want a `functions` prefix.

File: gitlab.yml (hunk 6 above)
Comment (alexellis) [reply gitlab.yml]: `https://gitlab.o6s.io/`

File: gitlab.yml [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    labels:
+      openfaas-cloud: "1"
+      role: openfaas-system
+    environment:
+      content_type: text/plain
+      validate_customers: true
+      validate_token: false
+      write_debug: true
+      read_debug: true
+      installation_tag: "openfaas-cloud"
```
Comment (alexellis) [inline gitlab.yml]: Make sure this defaults in the code for when it's commented out or deleted by mistake.

File: gitlab-event/handler.go [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+		if unmarshalErr != nil {
+			return fmt.Sprint("unable to unmarshal request into struct")
+		}
+
+		if readBool("validate_customers") {
+			var found bool
+			customersURL := os.Getenv("customers_url")
+			customers, getErr := getCustomers(customersURL)
+			if getErr != nil {
+				return getErr.Error()
```
[the remark this answers is not in the document]
Comment (alexellis) [reply gitlab-event/handler.go]: `unable to read customers from (put the value of customers_url) error: %s`

File: gitlab-event/handler.go [hunk 10]
```diff
@@ -0,0 +1,410 @@
...
+					Message: "Customer not found",
+					Owner:   eventInfo.UserUsername,
+					Source:  Source,
+				}
+				sdk.PostAudit(auditEvent)
+				return fmt.Sprintf("Customer: %s not found in CUSTOMERS file via %s", eventInfo.UserUsername, customersURL)
+			}
+		}
+
+		installed, err := InstalledApp(eventInfo.GitLabProject.ID, instance, apiToken, installationTag)
```
[the remark this answers is not in the document]
Comment (alexellis) [reply gitlab-event/handler.go]: I'd say adding a prefix of `check` is not needed for a bool function. `isAppInstalled` or `appInstalled` / `appEnabled` might be better. Any methods with capital letters are exported meaning you want to use them in other packages.. is that intentional?

File: gitlab-event/handler.go [hunk 11]
```diff
@@ -0,0 +1,410 @@
...
+		return val == "true" || val == "1"
+	}
+	return false
+}
+
+func tokenMatch(gitlabToken string, token string) bool {
+	return gitlabToken == token
+}
+
+func getCustomers(customerURL string) ([]string, error) {
```
Comment (alexellis) [inline gitlab-event/handler.go]: I was pretty sure we had this already in github-push or event, why are we duplicating the work?

File: gitlab.yml (hunk 8 above)
Comment (martindekov) [reply gitlab.yml]: Yeah currently it is

File: docs/README.md [hunk 12]
```diff
@@ -650,3 +650,128 @@ You will need to distribute or share pub-cert.pem so that people can use this wi
 ### Wildcard domains with of-router
 
 Coming soon
+
+# GitLab
```
[the remark this answers is not in the document]
Comment (martindekov) [reply docs/README.md]: Yeah good idea

File: gitlab-event/handler.go (hunk 11 above)
Comment (martindekov) [reply gitlab-event/handler.go]: Yeah I removed the customer validation from `gitlab-push` into the event function

File: gitlab-event/handler.go (hunk 4 above)
Comment (martindekov) [reply gitlab-event/handler.go]: Done

File: gitlab-event/handler.go [hunk 13]
```diff
@@ -0,0 +1,410 @@
...
+			return fmt.Sprintf("Function: `%s` deleted", eventInfo.Name)
+		}
+	}
+	return fmt.Sprintf("Message received with event: %s", eventName.Event)
+}
+
+func garbageCollect(garbageRequests []GarbageRequest) error {
+	client := http.Client{}
+
+	suffix := os.Getenv("dns_suffix")
```
[the remark this answers is not in the document]
Comment (martindekov) [reply gitlab-event/handler.go]: It is not merged into the code but It is working on kubernetes(tested on Bart instance) and swarm

File: gitlab-event/handler.go [hunk 14]
```diff
@@ -0,0 +1,410 @@
...
+		return fmt.Sprintf("%s: System Hook required cannot handle: %s", Source, eventHeader)
+	}
+
+	eventName := PureEvent{}
+	json.Unmarshal(req, &eventName)
+
+	if eventName.Event != "push" &&
+		eventName.Event != "project_update" &&
+		eventName.Event != "project_destroy" {
+
```
[the remark this answers is not in the document]
Comment (martindekov) [reply gitlab-event/handler.go]: Done, but naming is `checkSupportedEvents` saw the event name later

File: gitlab-event/handler.go [hunk 15]
```diff
@@ -0,0 +1,410 @@
...
+		return fmt.Sprintf("%s cannot handle event: %s", Source, eventName.Event)
+	}
+
+	if readBool("validate_token") {
+		tokenSecretKey, secretErr := sdk.ReadSecret("gitlab-webhook-secret")
+		if secretErr != nil {
+			return fmt.Sprintf("Unexpected error: %s", secretErr.Error())
+		}
+		if !tokenMatch(xGitlabToken, tokenSecretKey) {
+			return fmt.Sprintf("the request token and the existing tokens mismatch")
```
[the remark this answers is not in the document]
Comment (ivanayov) [reply gitlab-event/handler.go]: @martindekov I think the last suggestion is more descriptive. Can you please change?

File: gitlab-event/handler_test.go [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+func Test_getUser(t *testing.T) {
+	tests := []struct {
+		title             string
+		pathWithNamespace string
+		expectedName      string
+		expectedErr       error
+	}{
+		{
+			title:             "Error is nil and since the string is formated as expected",
+			pathWithNamespace: "exampleusername/exampleproject",
```
Comment (ivanayov) [inline gitlab-event/handler_test.go]: Let's add some more test cases, such as empty pathWithNamespace, etc.

File: gitlab-event/handler.go [hunk 17]
```diff
@@ -0,0 +1,410 @@
...
+	var installationTag string
+
+	if tag, exists := os.LookupEnv("installation_tag"); !exists {
+		installationTag = "openfaas-cloud"
+	} else {
+		installationTag = tag
+	}
+
+	switch eventName.Event {
+	case "push":
```
[the remark this answers is not in the document]
Comment (martindekov) [reply gitlab-event/handler.go]: They are in const now

File: gitlab-event/handler_test.go (hunk 16 above)
Comment (martindekov) [reply gitlab-event/handler_test.go]: Done

File: gitlab-event/handler.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (martindekov) [reply gitlab-event/handler.go]: changed to `appInstalled`
